src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/simple/SimpleJoinOrderingQueryOptimizer.java

```diff
@@ -41,14 +48,26 @@ public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws QueryOptimi
     }
 
     public PhysicalPlan optimizePlan( final PhysicalPlan plan ) throws QueryOptimizationException {
+        final PhysicalPlan[] optSubPlans = getOptimizedSubPlans(plan);
+
         if ( hasMultiwayJoinAsRoot(plan) ){
-            final PhysicalPlan[] subplans = getOptimizedSubPlans(plan);
-            return joinPlanOptimizer.determineJoinPlan(subplans);
+            return joinPlanOptimizer.determineJoinPlan(optSubPlans);
         }
-        else {
-        	// TODO incorrect!! There may be multiway joins inside the subplans of the given plan.
+        else if (hasNullaryOpAsRoot(plan)){
             return plan;
         }
+        else if ( hasUnaryOpAsRoot(plan) ){
+            final PhysicalPlan newChild = optimizePlan(optSubPlans[0]);
+            return new PhysicalPlanWithUnaryRootImpl((UnaryPhysicalOp) plan.getRootOperator(), newChild);
+        }
+        else if ( hasBinaryOpAsRoot(plan) ){
+            final PhysicalPlan newChild1 = optimizePlan(optSubPlans[0]);
+            final PhysicalPlan newChild2 = optimizePlan(optSubPlans[1]);
+            return new PhysicalPlanWithBinaryRootImpl((BinaryPhysicalOp) plan.getRootOperator(), newChild1, newChild2);
+        }
+        else {
```

Comment:
Before this final else-branch, there needs to be another else-if for the case of n-ary root operators other than multiway joins.